File: verenigingen/verenigingen_payments/page/sepa_mandate_diagnostics/sepa_mandate_diagnostics.js
frappe.pages['sepa_mandate_diagnostics'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('SEPA Mandate Diagnostics'),
		single_column: true
	});

	// Add refresh button
	page.add_inner_button(__('Refresh'), function() {
		load_diagnostics(page);
	});

	// Initial load
	load_diagnostics(page);
};

function load_diagnostics(page) {
	// Show loading indicator
	$(page.body).html('<div class="text-center margin-top"><p class="text-muted">' + __('Loading diagnostics...') + '</p></div>');

	// Fetch diagnostic data
	frappe.call({
		method: 'verenigingen.verenigingen_payments.page.sepa_mandate_diagnostics.sepa_mandate_diagnostics.get_mandate_issues',
		callback: function(r) {
			if (r.message) {
				render_diagnostics(page, r.message);
			}
		}
	});
}

function render_diagnostics(page, data) {
	const { issues, summary } = data;

	// Build summary card
	let html = `
		<div class="frappe-card margin-bottom">
			<div class="frappe-card-head">
				<strong>${__('Summary')}</strong>
			</div>
			<div class="frappe-card-body">
				<div class="row">
					<div class="col-md-4">
						<div class="text-center">
							<h2 class="${summary.total_issues > 0 ? 'text-danger' : 'text-success'}">${summary.total_issues}</h2>
							<p class="text-muted">${__('Total Issues')}</p>
						</div>
					</div>
					<div class="col-md-4">
						<div class="text-center">
							<h2>${summary.unique_members}</h2>
							<p class="text-muted">${__('Affected Members')}</p>
						</div>
					</div>
					<div class="col-md-4">
						<div class="text-center">
							<p class="text-muted">${__('Last Checked')}</p>
							<p><small>${frappe.datetime.str_to_user(summary.last_checked)}</small></p>
						</div>
					</div>
				</div>
			</div>
		</div>
	`;

	// Build issue cards
	Object.keys(issues).forEach(issue_key => {
		const issue = issues[issue_key];
		const severity_class = issue.severity === 'critical' ? 'danger' :
		                       issue.severity === 'high' ? 'orange' :
		                       issue.severity === 'medium' ? 'warning' : 'info';

		html += `
			<div class="frappe-card margin-bottom">
				<div class="frappe-card-head" data-toggle="collapse" data-target="#${issue_key}" style="cursor: pointer;">
					<span class="indicator ${severity_class}">${issue.title}</span>
					<span class="badge badge-${severity_class} pull-right">${issue.count}</span>
				</div>
				<div class="frappe-card-body">
					<p class="text-muted">${issue.description}</p>
					${issue.count > 0 && issue_key !== 'sepa_selected_no_mandate' ? `
						<button class="btn btn-sm btn-primary" onclick="fix_all_for_issue('${issue_key}')">${__('Fix All')}</button>
					` : ''}
					${issue_key === 'sepa_selected_no_mandate' && issue.count > 0 ? `
						<div class="alert alert-warning">
							<strong>${__('Manual Action Required')}</strong><br>
							${__('These members need SEPA mandates created. Auto-fix is not available. Please review each member and create mandates as needed.')}
						</div>
					` : ''}
				</div>
				<div id="${issue_key}" class="collapse ${issue.count > 0 && issue.count <= 20 ? 'in' : ''}">
					<div class="frappe-card-body" style="max-height: 400px; overflow-y: auto;">
						${render_issue_members(issue_key, issue.members)}
					</div>
				</div>
			</div>
		`;
	});

	$(page.body).html(html);
}

function render_issue_members(issue_type, members) {
	if (members.length === 0) {
		return '<p class="text-muted text-center">' + __('No issues found') + '</p>';
	}

	let html = '<table class="table table-bordered table-hover"><thead><tr>';

	// Dynamic headers based on issue type
	if (issue_type === 'sepa_selected_no_mandate') {
		html += '<th>' + __('Member') + '</th><th>' + __('Payment Method') + '</th><th>' + __('Total Mandates') + '</th><th>' + __('Active Mandates') + '</th><th>' + __('Actions') + '</th>';
	} else if (issue_type === 'missing_child_table_entries') {
		html += '<th>' + __('Member') + '</th><th>' + __('Mandate Count') + '</th><th>' + __('Mandate IDs') + '</th><th>' + __('Actions') + '</th>';
	} else if (issue_type === 'orphaned_child_table_entries') {
		html += '<th>' + __('Member') + '</th><th>' + __('Orphaned Mandate') + '</th><th>' + __('Reference') + '</th><th>' + __('Actions') + '</th>';
	} else if (issue_type === 'outdated_child_table_data') {
		html += '<th>' + __('Member') + '</th><th>' + __('Mandate') + '</th><th>' + __('Current Status') + '</th><th>' + __('Child Table Status') + '</th><th>' + __('Actions') + '</th>';
	} else if (issue_type === 'multiple_current_mandates') {
		html += '<th>' + __('Member') + '</th><th>' + __('Current Count') + '</th><th>' + __('Mandate IDs') + '</th><th>' + __('Actions') + '</th>';
	}

	html += '</tr></thead><tbody>';

	members.forEach(member => {
		html += '<tr>';

		if (issue_type === 'sepa_selected_no_mandate') {
			html += `
				<td><a href="/app/member/${member.member_id}">${member.full_name}</a></td>
				<td><span class="label label-info">${member.payment_method}</span></td>
				<td>${member.total_mandates}</td>
				<td><span class="label label-danger">${member.active_mandates}</span></td>
			`;
		} else if (issue_type === 'missing_child_table_entries') {
			html += `
				<td><a href="/app/member/${member.member_id}">${member.full_name}</a></td>
				<td>${member.mandate_count}</td>
				<td><small>${member.mandate_ids}</small></td>
			`;
		} else if (issue_type === 'orphaned_child_table_entries') {
			html += `
				<td><a href="/app/member/${member.member_id}">${member.full_name}</a></td>
				<td>${member.mandate_name}</td>
				<td>${member.mandate_reference}</td>
			`;
		} else if (issue_type === 'outdated_child_table_data') {
			html += `
				<td><a href="/app/member/${member.member_id}">${member.full_name}</a></td>
				<td>${member.mandate_id}</td>
				<td><span class="label label-info">${member.current_status}</span></td>
				<td><span class="label label-warning">${member.child_table_status}</span></td>
			`;
		} else if (issue_type === 'multiple_current_mandates') {
			html += `
				<td><a href="/app/member/${member.member_id}">${member.full_name}</a></td>
				<td>${member.current_count}</td>
				<td><small>${member.mandate_ids}</small></td>
			`;
		}

		html += `
			<td>
				<button class="btn btn-xs btn-success" onclick="fix_single_member('${member.member_id}')">${__('Fix')}</button>
			</td>
		</tr>`;
	});

	html += '</tbody></table>';

	return html;
}

function fix_single_member(member_id) {
	frappe.show_progress(__('Fixing...'), 0, 1, __('Fixing mandate issues for member'));

	frappe.call({
		method: 'verenigingen.verenigingen_payments.page.sepa_mandate_diagnostics.sepa_mandate_diagnostics.fix_member_mandate_issues',
		args: { member_id: member_id },
		callback: function(r) {
			frappe.hide_progress();

			if (r.message && r.message.success) {
				frappe.show_alert({
					message: __('Successfully fixed issues for {0}', [r.message.member_name]),
					indicator: 'green'
				});

				// Reload diagnostics
				load_diagnostics(cur_page.page);
			} else {
				frappe.show_alert({
					message: __('Failed to fix issues: {0}', [r.message.error]),
					indicator: 'red'
				});
			}
		}
	});
}

function fix_all_for_issue(issue_type) {
	frappe.confirm(
		__('This will fix all members with this issue type. Continue?'),
		function() {
			frappe.show_progress(__('Bulk Fix...'), 0, 1, __('Fixing mandate issues in batch'));

			frappe.call({
				method: 'verenigingen.verenigingen_payments.page.sepa_mandate_diagnostics.sepa_mandate_diagnostics.bulk_fix_mandate_issues',
				args: { issue_type: issue_type },
				callback: function(r) {
					frappe.hide_progress();

					if (r.message) {
						const msg = __('Fixed {0} of {1} members. {2} failures.', [
							r.message.success,
							r.message.total,
							r.message.failed
						]);

						frappe.show_alert({
							message: msg,
							indicator: r.message.failed > 0 ? 'orange' : 'green'
						});

						// Reload diagnostics
						load_diagnostics(cur_page.page);
					}
				}
			});
		}
	);
}

File: verenigingen/verenigingen_payments/page/sepa_mandate_diagnostics/sepa_mandate_diagnostics.py
"""
SEPA Mandate Diagnostics Page

Interactive diagnostic and repair tool for SEPA mandate synchronization issues.
Replaces blind nightly sync tasks with targeted problem detection and resolution.
"""

import frappe
from frappe import _

from verenigingen.utils.security.api_security_framework import OperationType, critical_api, standard_api


@frappe.whitelist()
@standard_api(operation_type=OperationType.READ)
def get_mandate_issues():
    """
    Get all SEPA mandate synchronization issues categorized by type.

    Returns:
        dict: Issue categories with counts and affected members
    """
    issues = {
        "sepa_selected_no_mandate": {
            "title": _("SEPA Payment Method Without Mandate"),
            "description": _(
                "Members have SEPA Direct Debit selected but lack active mandate or required banking data (IBAN, account holder name)"
            ),
            "severity": "critical",
            "count": 0,
            "members": [],
        },
        "missing_child_table_entries": {
            "title": _("Missing Child Table Entries"),
            "description": _("Members have SEPA mandates but no child table entries"),
            "severity": "high",
            "count": 0,
            "members": [],
        },
        "orphaned_child_table_entries": {
            "title": _("Orphaned Child Table Entries"),
            "description": _("Child table entries point to deleted mandates"),
            "severity": "medium",
            "count": 0,
            "members": [],
        },
        "outdated_child_table_data": {
            "title": _("Outdated Child Table Data"),
            "description": _("Child table data doesn't match current mandate status"),
            "severity": "low",
            "count": 0,
            "members": [],
        },
        "multiple_current_mandates": {
            "title": _("Multiple Current Mandates"),
            "description": _("Members marked as having multiple 'current' mandates"),
            "severity": "high",
            "count": 0,
            "members": [],
        },
        "mandate_member_data_mismatch": {
            "title": _("Mandate/Member Data Mismatch"),
            "description": _(
                "Active SEPA mandates where IBAN or account holder name differs from member record"
            ),
            "severity": "high",
            "count": 0,
            "members": [],
        },
    }

    # CRITICAL: Members with SEPA payment method but no active mandate
    sepa_no_mandate = frappe.db.sql(
        """
        SELECT
            m.name as member_id,
            m.full_name,
            m.payment_method,
            m.iban,
            m.bank_account_name,
            COUNT(sm.name) as total_mandates,
            COUNT(CASE WHEN sm.status = 'Active' AND sm.is_active = 1 THEN 1 END) as active_mandates,
            CASE
                WHEN m.iban IS NULL OR m.iban = '' THEN 'missing_iban'
                WHEN m.bank_account_name IS NULL OR m.bank_account_name = '' THEN 'missing_account_name'
                ELSE 'has_banking_data'
            END as banking_status
        FROM `tabMember` m
        LEFT JOIN `tabSEPA Mandate` sm ON sm.member = m.name
        WHERE m.payment_method = 'SEPA Direct Debit'
        GROUP BY m.name, m.full_name, m.payment_method, m.iban, m.bank_account_name
        HAVING active_mandates = 0
        """,
        as_dict=True,
    )
    issues["sepa_selected_no_mandate"]["count"] = len(sepa_no_mandate)
    issues["sepa_selected_no_mandate"]["members"] = sepa_no_mandate

    # Missing child table entries
    missing_entries = frappe.db.sql(
        """
        SELECT
            m.name as member_id,
            m.full_name,
            COUNT(sm.name) as mandate_count,
            GROUP_CONCAT(sm.mandate_id SEPARATOR ', ') as mandate_ids
        FROM `tabMember` m
        INNER JOIN `tabSEPA Mandate` sm ON sm.member = m.name
        LEFT JOIN `tabMember SEPA Mandate Link` sml ON sml.parent = m.name AND sml.sepa_mandate = sm.name
        WHERE sml.name IS NULL
        GROUP BY m.name, m.full_name
        """,
        as_dict=True,
    )
    issues["missing_child_table_entries"]["count"] = len(missing_entries)
    issues["missing_child_table_entries"]["members"] = missing_entries

    # Orphaned child table entries
    orphaned_entries = frappe.db.sql(
        """
        SELECT
            sml.parent as member_id,
            m.full_name,
            sml.sepa_mandate as mandate_name,
            sml.mandate_reference
        FROM `tabMember SEPA Mandate Link` sml
        INNER JOIN `tabMember` m ON m.name = sml.parent
        LEFT JOIN `tabSEPA Mandate` sm ON sm.name = sml.sepa_mandate
        WHERE sm.name IS NULL
        GROUP BY sml.parent, m.full_name, sml.sepa_mandate, sml.mandate_reference
        """,
        as_dict=True,
    )
    issues["orphaned_child_table_entries"]["count"] = len(orphaned_entries)
    issues["orphaned_child_table_entries"]["members"] = orphaned_entries

    # Outdated child table data
    outdated_data = frappe.db.sql(
        """
        SELECT
            m.name as member_id,
            m.full_name,
            sm.mandate_id,
            sm.status as current_status,
            sml.status as child_table_status,
            sm.sign_date as current_valid_from,
            sml.valid_from as child_table_valid_from
        FROM `tabMember` m
        INNER JOIN `tabSEPA Mandate` sm ON sm.member = m.name
        INNER JOIN `tabMember SEPA Mandate Link` sml ON sml.parent = m.name AND sml.sepa_mandate = sm.name
        WHERE sml.status != sm.status
           OR sml.mandate_reference != sm.mandate_id
           OR sml.valid_from != sm.sign_date
           OR sml.valid_until != sm.expiry_date
        """,
        as_dict=True,
    )
    issues["outdated_child_table_data"]["count"] = len(outdated_data)
    issues["outdated_child_table_data"]["members"] = outdated_data

    # Multiple current mandates
    multiple_current = frappe.db.sql(
        """
        SELECT
            m.name as member_id,
            m.full_name,
            COUNT(sml.name) as current_count,
            GROUP_CONCAT(sml.mandate_reference SEPARATOR ', ') as mandate_ids
        FROM `tabMember` m
        INNER JOIN `tabMember SEPA Mandate Link` sml ON sml.parent = m.name
        WHERE sml.is_current = 1
        GROUP BY m.name, m.full_name
        HAVING current_count > 1
        """,
        as_dict=True,
    )
    issues["multiple_current_mandates"]["count"] = len(multiple_current)
    issues["multiple_current_mandates"]["members"] = multiple_current

    # Mandate/Member data mismatch - IBAN or account holder name differs
    data_mismatch = frappe.db.sql(
        """
        SELECT
            m.name as member_id,
            m.full_name,
            m.iban as member_iban,
            m.bank_account_name as member_account_holder,
            sm.name as mandate_name,
            sm.mandate_id,
            sm.iban as mandate_iban,
            sm.account_holder_name as mandate_account_holder,
            CASE
                WHEN REPLACE(REPLACE(m.iban, ' ', ''), '-', '') != REPLACE(REPLACE(sm.iban, ' ', ''), '-', '')
                    AND (m.bank_account_name IS NULL OR m.bank_account_name = '' OR m.bank_account_name = sm.account_holder_name)
                    THEN 'iban_mismatch'
                WHEN m.bank_account_name IS NOT NULL AND m.bank_account_name != '' AND m.bank_account_name != sm.account_holder_name
                    AND REPLACE(REPLACE(m.iban, ' ', ''), '-', '') = REPLACE(REPLACE(sm.iban, ' ', ''), '-', '')
                    THEN 'holder_mismatch'
                ELSE 'both_mismatch'
            END as mismatch_type
        FROM `tabMember` m
        INNER JOIN `tabSEPA Mandate` sm ON sm.member = m.name
        WHERE sm.status = 'Active'
          AND sm.is_active = 1
          AND (
              -- IBAN mismatch (normalized comparison without spaces/dashes)
              REPLACE(REPLACE(m.iban, ' ', ''), '-', '') != REPLACE(REPLACE(sm.iban, ' ', ''), '-', '')
              -- OR account holder name mismatch (when member has one set)
              OR (
                  m.bank_account_name IS NOT NULL
                  AND m.bank_account_name != ''
                  AND m.bank_account_name != sm.account_holder_name
              )
          )
        """,
        as_dict=True,
    )
    issues["mandate_member_data_mismatch"]["count"] = len(data_mismatch)
    issues["mandate_member_data_mismatch"]["members"] = data_mismatch

    # Calculate totals
    total_issues = sum(issue["count"] for issue in issues.values())
    unique_members = len(
        set(member.get("member_id") for issue in issues.values() for member in issue["members"])
    )

    return {
        "issues": issues,
        "summary": {
            "total_issues": total_issues,
            "unique_members": unique_members,
            "last_checked": frappe.utils.now(),
        },
    }


@frappe.whitelist()
@critical_api(operation_type=OperationType.FINANCIAL)
def fix_member_mandate_issues(member_id, issue_types=None):
    """
    Fix SEPA mandate issues for a specific member.

    Args:
        member_id: Member document name
        issue_types: List of issue types to fix (None = all)

    Returns:
        dict: Success status and details
    """
    try:
        member = frappe.get_doc("Member", member_id)

        # Use the existing refresh method
        result = member.refresh_sepa_mandates_table()

        return {"success": True, "member": member_id, "member_name": member.full_name, "details": result}

    except Exception as e:
        frappe.log_error(f"Failed to fix mandate issues for {member_id}: {str(e)}", "SEPA Mandate Fix Error")
        return {"success": False, "member": member_id, "error": str(e)}


@frappe.whitelist()
@critical_api(operation_type=OperationType.FINANCIAL)
def bulk_fix_mandate_issues(issue_type=None, member_ids=None):
    """
    Fix SEPA mandate issues for multiple members.

    Args:
        issue_type: Specific issue type to fix (optional)
        member_ids: List of member IDs (optional, defaults to all with issues)

    Returns:
        dict: Batch operation results
    """
    results = {"total": 0, "success": 0, "failed": 0, "errors": []}

    # If no member IDs provided, get all members with issues
    if not member_ids:
        issues = get_mandate_issues()

        if issue_type and issue_type in issues["issues"]:
            # Fix specific issue type
            member_ids = [m["member_id"] for m in issues["issues"][issue_type]["members"]]
        else:
            # Fix all issues
            all_members = set()
            for issue_data in issues["issues"].values():
                all_members.update(m["member_id"] for m in issue_data["members"])
            member_ids = list(all_members)

    results["total"] = len(member_ids)

    # Process each member
    for member_id in member_ids:
        fix_result = fix_member_mandate_issues(member_id)

        if fix_result["success"]:
            results["success"] += 1
        else:
            results["failed"] += 1
            results["errors"].append({"member": member_id, "error": fix_result.get("error", "Unknown error")})

    # Log summary if there were errors
    if results["errors"]:
        error_summary = f"Failed to fix {results['failed']} of {results['total']} members:\n\n"
        for err in results["errors"][:20]:
            error_summary += f"- {err['member']}: {err['error']}\n"
        if len(results["errors"]) > 20:
            error_summary += f"\n... and {len(results['errors']) - 20} more errors"

        frappe.log_error(error_summary, "SEPA Mandate Bulk Fix Errors")

    return results
